README.md
Geobricks GIS Raster
====================

Geobricks library to process raster layers.


CHANGES.txt
v0.0.1, 26-11-2014 -- Initial release.
v0.1.1, 01-12-2014 -- Updated setup.py
v0.1.2, 10-12-2014 -- Updated setup.py
v0.1.2, 10-12-2014 -- Updated filesystem
v0.1.3, 19-12-2014 -- Added GebobricksCore dependency.
v0.1.4, 19-12-2014 -- Changed to GeobricksCommon dependency.
v0.1.5, 13-01-2015 -- Changed dependencies.
v0.1.6, 11-02-2015 -- Added osgeo dependency.
v0.1.7, 11-02-2015 -- Added gdal dependency.
v0.1.8, 11-02-2015 -- Added gdal dependency.
v0.1.9, 11-02-2015 -- Removed gdal dependency.
v0.1.10, 16-03-2015 -- added proj4 to epsg dependency.
v0.1.11, 16-03-2015 -- typo in config.
v0.1.12, 16-03-2015 -- typo in config.


test/raster_test.py
import unittest
from geobricks_gis_raster.core.raster import get_location_values, get_srid, get_authority, get_nodata_value, get_histogram, get_descriptive_statistics, get_statistics


class GeobricksTest(unittest.TestCase):

    def test_get_nodata(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        result = get_nodata_value(path)
        self.assertEqual(result, '8.99999982852e+20')

    def test_get_statistics(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        result = get_statistics(path)
        self.assertEqual(result, {'hist': [{'band': 1, 'buckets': 256, 'values': [2038201, 37421, 20886, 13506, 9780, 7553, 6260, 5304, 4485, 4115, 3789, 3437, 3115, 3324, 2229, 1952, 1919, 1808, 1745, 1584, 1486, 1468, 1299, 1146, 1144, 1139, 1053, 961, 924, 879, 794, 818, 747, 788, 724, 709, 677, 643, 595, 626, 552, 581, 602, 494, 507, 420, 463, 466, 418, 410, 432, 364, 368, 362, 424, 315, 319, 260, 275, 235, 250, 232, 211, 205, 230, 207, 200, 228, 212, 209, 202, 174, 213, 238, 153, 165, 134, 158, 161, 122, 116, 112, 103, 119, 117, 93, 97, 106, 95, 78, 82, 90, 96, 86, 69, 105, 91, 76, 83, 76, 78, 88, 92, 59, 76, 90, 61, 69, 96, 81, 60, 64, 52, 70, 56, 52, 72, 62, 89, 58, 82, 69, 63, 64, 78, 42, 62, 37, 40, 47, 59, 41, 39, 48, 33, 52, 38, 35, 58, 32, 22, 26, 19, 26, 25, 12, 20, 12, 13, 18, 21, 41, 17, 10, 10, 9, 17, 9, 30, 10, 6, 7, 11, 4, 7, 3, 7, 7, 4, 1, 8, 5, 6, 6, 17, 10, 6, 4, 6, 6, 9, 6, 8, 8, 4, 6, 10, 5, 7, 7, 24, 4, 0, 3, 18, 1, 1, 1, 3, 1, 2, 1, 1, 2, 1, 2, 5, 0, 0, 5, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1], 'max': 1.9333192110061646, 'min': 0.0}], 'stats': [{'band': 1, 'max': 1.9333192110061646, 'mean': 0.008779648138849024, 'sd': 0.056981237827174326, 'min': 0.0}]})

    def test_get_histogram(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        result = get_histogram(path)
        self.assertEqual(result, [{'band': 1, 'buckets': 256, 'values': [2038201, 37421, 20886, 13506, 9780, 7553, 6260, 5304, 4485, 4115, 3789, 3437, 3115, 3324, 2229, 1952, 1919, 1808, 1745, 1584, 1486, 1468, 1299, 1146, 1144, 1139, 1053, 961, 924, 879, 794, 818, 747, 788, 724, 709, 677, 643, 595, 626, 552, 581, 602, 494, 507, 420, 463, 466, 418, 410, 432, 364, 368, 362, 424, 315, 319, 260, 275, 235, 250, 232, 211, 205, 230, 207, 200, 228, 212, 209, 202, 174, 213, 238, 153, 165, 134, 158, 161, 122, 116, 112, 103, 119, 117, 93, 97, 106, 95, 78, 82, 90, 96, 86, 69, 105, 91, 76, 83, 76, 78, 88, 92, 59, 76, 90, 61, 69, 96, 81, 60, 64, 52, 70, 56, 52, 72, 62, 89, 58, 82, 69, 63, 64, 78, 42, 62, 37, 40, 47, 59, 41, 39, 48, 33, 52, 38, 35, 58, 32, 22, 26, 19, 26, 25, 12, 20, 12, 13, 18, 21, 41, 17, 10, 10, 9, 17, 9, 30, 10, 6, 7, 11, 4, 7, 3, 7, 7, 4, 1, 8, 5, 6, 6, 17, 10, 6, 4, 6, 6, 9, 6, 8, 8, 4, 6, 10, 5, 7, 7, 24, 4, 0, 3, 18, 1, 1, 1, 3, 1, 2, 1, 1, 2, 1, 2, 5, 0, 0, 5, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1], 'max': 1.9333192110061646, 'min': 0.0}])

    def test_get_descriptive_statistics(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        result = get_descriptive_statistics(path)
        self.assertEqual(result, [{'band': 1, 'max': 1.9333192110061646, 'mean': 0.008779648138849024, 'sd': 0.056981237827174326, 'min': 0.0}])

    def test_get_srid(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        result = get_srid(path)
        self.assertEqual(result, '4326')

    def test_get_authority(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        result = get_authority(path)
        self.assertEqual(result, 'epsg:4326')

    def test_get_location_values(self):
        path = ["../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"]
        result = get_location_values(path, 45.1709201, 7.985972)
        self.assertEqual(result, ['0.1346156001091'])


if __name__ == '__main__':
    suite = unittest.TestLoader().loadTestsFromTestCase(GeobricksTest)
    unittest.TextTestRunner(verbosity=2).run(suite)



geobricks_gis_raster_test/raster_test.py
import unittest
import os
from geobricks_gis_raster.core.raster import get_location_values, get_srid, get_authority, get_nodata_value, get_histogram, get_descriptive_statistics, get_statistics


class GeobricksTest(unittest.TestCase):

    def test_get_nodata(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = os.path.normpath(os.path.join(os.path.dirname(__file__), path))
        result = get_nodata_value(path)
        self.assertEqual(result, '8.99999982852e+20')

    def test_get_statistics(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = os.path.normpath(os.path.join(os.path.dirname(__file__), path))
        result = get_statistics(path)
        self.assertEqual(result, {'hist': [{'band': 1, 'buckets': 256, 'values': [2038201, 37421, 20886, 13506, 9780, 7553, 6260, 5304, 4485, 4115, 3789, 3437, 3115, 3324, 2229, 1952, 1919, 1808, 1745, 1584, 1486, 1468, 1299, 1146, 1144, 1139, 1053, 961, 924, 879, 794, 818, 747, 788, 724, 709, 677, 643, 595, 626, 552, 581, 602, 494, 507, 420, 463, 466, 418, 410, 432, 364, 368, 362, 424, 315, 319, 260, 275, 235, 250, 232, 211, 205, 230, 207, 200, 228, 212, 209, 202, 174, 213, 238, 153, 165, 134, 158, 161, 122, 116, 112, 103, 119, 117, 93, 97, 106, 95, 78, 82, 90, 96, 86, 69, 105, 91, 76, 83, 76, 78, 88, 92, 59, 76, 90, 61, 69, 96, 81, 60, 64, 52, 70, 56, 52, 72, 62, 89, 58, 82, 69, 63, 64, 78, 42, 62, 37, 40, 47, 59, 41, 39, 48, 33, 52, 38, 35, 58, 32, 22, 26, 19, 26, 25, 12, 20, 12, 13, 18, 21, 41, 17, 10, 10, 9, 17, 9, 30, 10, 6, 7, 11, 4, 7, 3, 7, 7, 4, 1, 8, 5, 6, 6, 17, 10, 6, 4, 6, 6, 9, 6, 8, 8, 4, 6, 10, 5, 7, 7, 24, 4, 0, 3, 18, 1, 1, 1, 3, 1, 2, 1, 1, 2, 1, 2, 5, 0, 0, 5, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1], 'max': 1.9333192110061646, 'min': 0.0}], 'stats': [{'band': 1, 'max': 1.9333192110061646, 'mean': 0.008779648138849024, 'sd': 0.056981237827174326, 'min': 0.0}]})

    def test_get_histogram(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = os.path.normpath(os.path.join(os.path.dirname(__file__), path))
        result = get_histogram(path)
        self.assertEqual(result, [{'band': 1, 'buckets': 256, 'values': [2038201, 37421, 20886, 13506, 9780, 7553, 6260, 5304, 4485, 4115, 3789, 3437, 3115, 3324, 2229, 1952, 1919, 1808, 1745, 1584, 1486, 1468, 1299, 1146, 1144, 1139, 1053, 961, 924, 879, 794, 818, 747, 788, 724, 709, 677, 643, 595, 626, 552, 581, 602, 494, 507, 420, 463, 466, 418, 410, 432, 364, 368, 362, 424, 315, 319, 260, 275, 235, 250, 232, 211, 205, 230, 207, 200, 228, 212, 209, 202, 174, 213, 238, 153, 165, 134, 158, 161, 122, 116, 112, 103, 119, 117, 93, 97, 106, 95, 78, 82, 90, 96, 86, 69, 105, 91, 76, 83, 76, 78, 88, 92, 59, 76, 90, 61, 69, 96, 81, 60, 64, 52, 70, 56, 52, 72, 62, 89, 58, 82, 69, 63, 64, 78, 42, 62, 37, 40, 47, 59, 41, 39, 48, 33, 52, 38, 35, 58, 32, 22, 26, 19, 26, 25, 12, 20, 12, 13, 18, 21, 41, 17, 10, 10, 9, 17, 9, 30, 10, 6, 7, 11, 4, 7, 3, 7, 7, 4, 1, 8, 5, 6, 6, 17, 10, 6, 4, 6, 6, 9, 6, 8, 8, 4, 6, 10, 5, 7, 7, 24, 4, 0, 3, 18, 1, 1, 1, 3, 1, 2, 1, 1, 2, 1, 2, 5, 0, 0, 5, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1], 'max': 1.9333192110061646, 'min': 0.0}])

    def test_get_descriptive_statistics(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = os.path.normpath(os.path.join(os.path.dirname(__file__), path))
        result = get_descriptive_statistics(path)
        self.assertEqual(result, [{'band': 1, 'max': 1.9333192110061646, 'mean': 0.008779648138849024, 'sd': 0.056981237827174326, 'min': 0.0}])

    def test_get_srid(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = os.path.normpath(os.path.join(os.path.dirname(__file__), path))
        result = get_srid(path)
        self.assertEqual(result, '4326')

    def test_get_authority(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = os.path.normpath(os.path.join(os.path.dirname(__file__), path))
        result = get_authority(path)
        self.assertEqual(result, 'epsg:4326')

    def test_get_location_values(self):
        path = "../test_data/storage/raster/rice_area_4326/rice_area_4326.geotiff"
        path = [os.path.normpath(os.path.join(os.path.dirname(__file__), path))]
        result = get_location_values(path, 45.1709201, 7.985972)
        self.assertEqual(result, ['0.1346156001091'])


def run_test():
    suite = unittest.TestLoader().loadTestsFromTestCase(GeobricksTest)
    unittest.TextTestRunner(verbosity=2).run(suite)

if __name__ == '__main__':
    run_test()






examples/get_raster_metadata.py
from geobricks_gis_raster.core import raster

try:
    print "test"
    path = "/home/vortex/Desktop/LAYERS/GHG_13_NOVEMEBRE/MCD12Q1/processed_2001/4326/final_4326.tiff"
    path = "/home/vortex/Desktop/LAYERS/GHG_13_NOVEMEBRE/MCD12Q1/processed_2001/0.hdf"
    path = "/home/vortex/Desktop/LAYERS/GHG_13_NOVEMEBRE/cultivation_organic_soils/3857/cultivation_organic_soils_croplands_3857.tif"
    #path = "/home/vortex/Desktop/LAYERS/GHG_13_NOVEMEBRE/Climate_Zones_processed/JRC_CLIMATE_ZONE_0.25deg_3857.tif"
    #path = "/home/vortex/Desktop/LAYERS/GHG_13_NOVEMEBRE/Climate_Zones_processed/JRC_CLIMATE_ZONE_0.25deg_4326.tif"

    # multiband
    # path = "/home/vortex/Desktop/LAYERS/MODIS_5600/MODIS_03_2014.hdf"
    print path
    #result = raster.get_nodata_value(path)
    result = raster.get_authority(path)

    print result
except Exception, e:
    print(e)




examples/export_raster_example.py
from geobricks_gis_raster.core.raster import crop_by_vector_database, get_statistics, get_authority
from geobricks_gis_raster.config.config import config
from geobricks_dbms.core.dbms_postgresql import DBMSPostgreSQL
import rasterio

config = config["settings"]

# request (to be passed to the distribution service)
r = {
    "raster": {
        "uids":["fenix:gfed4_burnedarea_lc_12_2012"]
    },
    "vector": {
        "query_extent": "SELECT ST_AsGeoJSON(ST_Transform(ST_SetSRID(ST_Extent(geom), 3857), {{SRID}})) FROM {{SCHEMA}}.gaul0_2014_2013_3857 WHERE adm0_code IN ('122')",
        "query_layer":  "SELECT * FROM {{SCHEMA}}.gaul0_2014_2013_3857 WHERE adm0_name IN ('Portugal')"
    }
}


# DB Instance
db_spatial = DBMSPostgreSQL(config["db"]["spatial"])

# queries
raster_path = "../test_data/MODIS/MOD13A2/MOD13A2_3857.tif"
query_extent = "SELECT ST_AsGeoJSON(ST_Transform(ST_SetSRID(ST_Extent(geom), 3857), 3857)) FROM spatial.gaul0_2014_2013_3857 WHERE adm0_name IN ('Portugal')"
query_layer = "SELECT * FROM spatial.gaul0_2014_2013_3857 WHERE adm0_name IN ('Portugal')"

# cutting the layer
result = crop_by_vector_database(raster_path, db_spatial, query_extent, query_layer)
print result
print get_statistics(result)
print get_authority(result)



geobricks_gis_raster/config/config.py
import logging

config = {

    "settings": {

        # To be used by Flask: DEVELOPMENT ONLY
        "debug": True,

        # Flask host: DEVELOPMENT ONLY
        "host": "localhost",

        # Flask port: DEVELOPMENT ONLY
        "port": 5904,

        # Logging configurations
        "logging": {
            "level": logging.INFO,
            "format": "%(asctime)s | %(levelname)-8s | %(name)-20s | Line: %(lineno)-5d | %(message)s",
            "datefmt": "%d-%m-%Y | %H:%M:%s"
        },

        # Databases
        "db": {
            # Spatial Database
            "spatial": {
                # default_db will search in the dbs["database"] as default option
                "dbname": "fenix",
                "host": "EXLPRFAOSTAT1.ext.fao.org",
                "port": "5432",
                "username": "fenix",
                "password": "Qwaszx",
                "schema": "spatial"
            }
        },

        # Databases
        # "db": {
        #     # Spatial Database
        #     "spatial": {
        #         # default_db will search in the dbs["database"] as default option
        #         "dbname": "dbname",
        #         "host": "localhost",
        #         "port": "5432",
        #         "username": "user",
        #         "password": "password",
        #         "schema": "schema"
        #     }
        # }
    }
}


geobricks_gis_raster/utils/filesystem.py
import os
import tempfile
import uuid
from geobricks_gis_raster.config.config import config

# temporary folder
try:
    folder_tmp_default = config["settings"]["folders"]["tmp"]
except Exception:
    folder_tmp_default = tempfile.gettempdir()

try:
    workspace_layer_separator = config["settings"]["folders"]["workspace_layer_separator"]
except Exception, e:
    workspace_layer_separator = "@"


def create_tmp_filename(extension='', filename='',  subfolder='', add_uuid=True, folder_tmp=folder_tmp_default):
    """
    :param extension: i.e. "tif"
    :param subfolder: "modis_folder"
    :param filename: "modis"
    :param add_uuid: add or not a uuid to the file
    :param folder_tmp: if not specified it takes the default tmp folder of the os.
    :return: a path to a tmp file
    """

    """
    Create the path for a tmp file and filename

    @type path: string
    @param path: path from the tmp folder
    @type extension: extension
    @param extension: i.e. .geotiff
    """
    if extension != '' and "." not in extension: extension = "." + extension
    folder_path = os.path.join(folder_tmp, subfolder)
    if subfolder is not '':
        try:
            os.makedirs(folder_path)
        except Exception, e:
            # log.warn(e)
            pass
    file_path = os.path.join(folder_path, filename)
    if add_uuid:
        return (file_path + str(uuid.uuid4()) + extension).encode('utf-8')
    else:
        return (file_path + extension).encode('utf-8')


# TODO: move it to the data manager?
def get_raster_path_by_uid(uid, ext=".geotiff"):
    l = uid.split(workspace_layer_separator) if workspace_layer_separator in uid else uid.split(":")
    return os.path.join(config["settings"]["folders"]["geoserver_datadir"], "data",  l[0], l[1], l[1] + ext);


def get_raster_path_by_ftp_uid(uid, ext=".geotiff"):
    l = uid.split(workspace_layer_separator) if workspace_layer_separator in uid else uid.split(":")
    return os.path.join(config["settings"]["folders"]["ftp"], l[0], l[1], l[1] + ext)


geobricks_gis_raster/core/raster_mapalgebra.py
import numpy as np
import gdal
from osr import SpatialReference
from geobricks_common.core.log import logger
from geobricks_common.core.filesystem import create_tmp_filename

log = logger(__file__)


def mask(raster_path1, raster_path2, min1=0, max1=None, min2=0, max2=None, band1=1, band2=1):
    ds1 = gdal.Open(raster_path1)
    ds2 = gdal.Open(raster_path2)
    rows1 = ds1.RasterYSize
    cols1 = ds1.RasterXSize
    rows2 = ds2.RasterYSize
    cols2 = ds2.RasterXSize

    log.info("Scatter Processing")
    if cols1 != cols2 or rows1 != rows2:
        log.error("The rasters cannot be processed because they have different dimensions")
        log.error("%sx%s %sx%s" % (rows1, cols1, rows2, cols2))
        raise Exception("The rasters cannot be processed because they have different dimensions")

    band1 = ds1.GetRasterBand(band1)
    array1 = np.array(band1.ReadAsArray()).flatten()
    #array1 = np.array(band1.ReadAsArray())

    nodata1 = band1.GetNoDataValue()

    band2 = ds2.GetRasterBand(band2)
    array2 = np.array(band2.ReadAsArray()).flatten()
    #array2 = np.array(band2.ReadAsArray())
    nodata2 = band2.GetNoDataValue()

    # merge the layers
    path = _merge_layers(cols1, rows1, ds1.GetGeoTransform(), ds1.GetProjection(), array1, array2, min1, max1, min2, max2, nodata1, nodata2)   # is it useful to remove them fro the memory?
    del ds1, ds2, array1, array2
    return path


def _merge_layers(rows, cols, geotransform, spatialreference, array1, array2, min1, max1, min2, max2, nodata1=None, nodata2=None):
    path = create_tmp_filename('', ".tif")

    # find the indexes of the arrays
    index1 = (array1 > min1) & (array1 <= max1) & (array1 != nodata1)
    index2 = (array2 > min2) & (array2 <= max2) & (array2 != nodata2)

    # merge array indexes
    compound_index = index1 & index2
    del index1, index2

    # create a new raster
    output_raster = gdal.GetDriverByName('GTiff').Create(path, rows, cols, 1, gdal.GDT_Int16)  # Open the file
    output_raster.SetGeoTransform(geotransform)
    srs = SpatialReference(wkt=spatialreference)
    output_raster.SetProjection(srs.ExportToWkt())
    # create raster from the compound_index of the two rasters
    # TODO: the reshape slows the operation, use matrixes
    output_raster.GetRasterBand(1).WriteArray(compound_index.reshape(cols, rows))
    return path

geobricks_gis_raster/core/raster.py
import gdal
import os
import subprocess
import copy
import math
import json
import rasterio
from geobricks_common.core.log import logger
from geobricks_common.core.filesystem import create_tmp_filename
from geobricks_proj4_to_epsg.core.proj4_to_epsg import get_epsg_code_from_proj4

log = logger(__file__)

# example of statistics
stats_config = {
    "descriptive_stats": {
        "force": True
    },
    "histogram": {
        "buckets": 256,
        "include_out_of_range": 0,
        "force": True
    }
}


def crop_raster_on_vector_bbox_and_postgis_db(input_file, db_connection_string, query, minlon, minlat, maxlon, maxlat, srcnodata=None, dstnodata=None):
    if srcnodata == None:
        srcnodata = get_nodata_value(input_file)
    if dstnodata == None:
        dstnodata = srcnodata

    output_bbox = crop_raster_with_bounding_box(input_file, minlon, minlat, maxlon, maxlat, srcnodata)
    output_path = crop_by_vector_from_db(output_bbox, db_connection_string, query, srcnodata, dstnodata)
    os.remove(output_bbox)
    return output_path


def crop_raster_with_bounding_box(input_file, minlon, minlat, maxlon, maxlat, srcnodata=None):
    if srcnodata == None:
        srcnodata = get_nodata_value(input_file)

    log.info("crop_raster_with_bounding_box")
    output_file = create_tmp_filename('.tif', 'gdal_translate_by_bbox')
    args = [
        "gdal_translate",
        "-a_nodata", str(srcnodata),
        "-projwin",
        str(minlat),
        str(minlon),
        str(maxlat),
        str(maxlon),
        input_file,
        output_file
    ]
    try:
        log.info(args)
        proc = subprocess.call(args, stdout=subprocess.PIPE, stderr=None)
        # proc = subprocess.check_output(args)
    except Exception, e:
        raise Exception(e)
    return output_file


def crop_by_vector_from_db(input_file, db_connection_string, query, srcnodata='nodata', dstnodata='nodata'):
    log.info(query)
    output_file_gdal_warp = create_tmp_filename('.tif', 'gdal_warp')
    output_file = create_tmp_filename('.tif', 'output')
    log.info(input_file)

    # crop the layer on cutline
    args = [
        'gdalwarp',
        "-q",
        "-multi",
        "-of", "GTiff",
        "-cutline", db_connection_string,
        "-csql", query,
        "-srcnodata", str(srcnodata),
        "-dstnodata", str(dstnodata),
        # -crop_to_cutline is needed otherwise the layer is not cropped
        # TODO: resolve shifting problem
        # "-crop_to_cutline",
        # "-dstalpha",
        input_file,
        output_file_gdal_warp
    ]
    try:
        log.info(args)
        #TODO: handle subprocess Error (like that is not taken)
        output = subprocess.check_output(args)
        # stdout_value, error = proc.communicate()
        log.info(output)
    except Exception, e:
        raise Exception(e)

        # TODO: is it useful the third opetation?
    args = [
        'gdal_translate',
        "-co", "COMPRESS=DEFLATE",
        "-a_nodata", str(dstnodata),
        output_file_gdal_warp,
        output_file
    ]
    try:
        log.info(" ".join(args))
        #TODO: handle subprocess Error (like that is not taken)
        proc = subprocess.call(args, stdout=subprocess.PIPE, stderr=None)
    except:
        stdout_value = proc.communicate()[0]
        raise Exception(stdout_value)

    #os.remove(output_file_gdal_warp)
    if os.path.isfile(output_file):
        return output_file
    return None



############### OLD METHODS
# TODO: remove the db_spatial from here
def crop_by_vector_database(raster_path, db_spatial, query_extent, query_layer):
    # TODO: make it better
    geom = json.dumps(db_spatial.query(query_extent))
    g = json.loads(geom)
    #log.info(g)
    obj = g[0][0]
    #log.info(obj)
    obj = json.loads(obj)
    # TODO: this is hardcoded because the returning bbox is different from the one used by GDAL processing
    #log.info(obj["coordinates"])
    minlat = obj["coordinates"][0][0][0]
    minlon = obj["coordinates"][0][1][1]
    maxlat = obj["coordinates"][0][2][0]
    maxlon = obj["coordinates"][0][0][1]
    db_connection_string = db_spatial.get_connection_string(True)
    srcnodatavalue = get_nodata_value(raster_path)
    return _crop_by_vector_database(raster_path, query_layer, db_connection_string, minlat, minlon, maxlat, maxlon, srcnodatavalue, srcnodatavalue)


# TODO: instead of the connection string pass the geometry
def _crop_by_vector_database(input_file, query, db_connection_string, minlat, minlon, maxlat, maxlon, srcnodata='nodata', dstnodata='nodata'):
    log.info(query)
    output_file_gdal_translate = create_tmp_filename('.tif', 'gdal_translate')
    output_file_gdal_warp = create_tmp_filename('.tif', 'gdal_warp')
    output_file = create_tmp_filename('.tif', 'output')
    args = [
        'gdal_translate',
        '-projwin',
        str(minlat),
        str(minlon),
        str(maxlat),
        str(maxlon),
        input_file,
        output_file_gdal_translate
    ]
    try:
        log.info(args)
        #TODO: handle subprocess Error (like that is not taken)
        proc = subprocess.call(args, stdout=subprocess.PIPE, stderr=None)
    except:
        stdout_value = proc.communicate()[0]
        raise Exception(stdout_value)

    args = [
        'gdalwarp',
        "-q",
        "-multi",
        "-of", "GTiff",
        "-cutline", db_connection_string,
        "-csql", query,
        "-srcnodata", str(srcnodata),
        "-dstnodata", str(dstnodata),
        # -crop_to_cutline is needed otherwise the layer is not cropped
        # TODO: resolve shifting problem
        # "-crop_to_cutline",
        # "-dstalpha",
        output_file_gdal_translate,
        output_file_gdal_warp
    ]
    try:
        #log.info(args)
        #TODO: handle subprocess Error (like that is not taken)
        proc = subprocess.call(args, stdout=subprocess.PIPE, stderr=None)
    except:
        stdout_value = proc.communicate()[0]
        raise Exception(stdout_value)

    # TODO: is it useful the third opetation?
    args = [
        'gdal_translate',
        "-co", "COMPRESS=DEFLATE",
        "-a_nodata", str(dstnodata),
        output_file_gdal_warp,
        output_file
    ]
    try:
        log.info(args)
        #TODO: handle subprocess Error (like that is not taken)
        proc = subprocess.call(args, stdout=subprocess.PIPE, stderr=None)
    except:
        stdout_value = proc.communicate()[0]
        raise Exception(stdout_value)

    os.remove(output_file_gdal_warp)
    os.remove(output_file_gdal_translate)

    if os.path.isfile(output_file):
        return output_file
    return None


def get_statistics(input_file, config=stats_config):
    """
    :param input_file: file to be processed
    :param config: json config file to be passed
    :return: computed statistics
    """
    #log.info("get_statistics: %s" % input_file)

    if config is None:
        config = copy.deepcopy(stats_config)

    stats = {}
    try:
        if os.path.isfile(input_file):
            ds = gdal.Open(input_file)
            if "descriptive_stats" in config:
                stats["stats"] = _get_descriptive_statistics(ds, config["descriptive_stats"])
            if "histogram" in config:
                stats["hist"] = _get_histogram(ds, config["histogram"])
        else:
            raise Exception("Exceptiuon")
    except Exception, e:
        raise Exception(e)
    return stats


def get_descriptive_statistics(input_file, config=stats_config["descriptive_stats"]):
    """
    :param input_file: file to be processed
    :param config: json config file to be passed
    :return: return and array with the min, max, mean, sd statistics per band i.e. [{"band": 1, "max": 549.0, "mean": 2.8398871527778, "sd": 17.103028971129, "min": 0.0}]
    """
    try:
        if os.path.isfile(input_file):
            ds = gdal.Open(input_file)
            return _get_descriptive_statistics(ds, config)
        else:
            # TODO exception
            raise Exception("Exceptiuon")
    except Exception, e:
        raise Exception(e)


def get_histogram(input_file, config=stats_config["histogram"]):
    """
    :param input_file: file to be processed
    :type string
    :param config: json config file to be passed
    :type json
    :return: return and array with the min, max, mean, sd statistics per band i.e. [{"band": 1, "buckets": 256, "values": [43256, 357, ...], "max": 998.0, "min": 0.0}]
    """
    try:
        if os.path.isfile(input_file):
            ds = gdal.Open(input_file)
            return _get_histogram(ds, config)
        else:
            # TODO exception
            raise Exception("Exceptiuon")
    except Exception, e:
        raise Exception(e)



def get_location_values(input_files, lat, lon, band=None):
    """
    Get the value of a (lat, lon) location

    # TODO:
    1) pass a json, instead of [files] pass file and id
    2) pass as well the projection used i.e. EPSG:4326
    3) for now it's used x, y as lat lon (it's not used the projection)

    :param input_files: files to be processed
    :type array
    :param lat: x value (for now it's used LatLon)
    :type float
    :param lon: y value (for now it's used LatLon)
    :type float
    :param band: band default=None (not yet used)
    :return: and array with the values of the (lat, lon) location
    """
    values = []
    for input_file in input_files:
        values.append(_location_value(input_file, lat, lon, band))
    return values


def _location_value(input_file, lat, lon, band=None):
    """
     Get the value of a (lat, lon) location
    :param input_file: file to be processed
    :type string
    :param x: x value
    :type float
    :param y: y value
    :type float
    :param band: band default=None (not yet used)
    :return: and array with the values of the (lat, lon) location
    """
    # TODO: check with -wgs84 values instead of -geoloc that is the reference system of the image
    cmd = "gdallocationinfo -valonly " + input_file + " -wgs84 " + str(lon) + " " + str(lat)

    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
    output, error = process.communicate()
    return output.strip()


def _get_descriptive_statistics(ds, config):
    # variables
    force = True if "force" not in config else bool(config["force"])

    # stats
    stats = []
    for band in range(ds.RasterCount):
        band += 1
        srcband = ds.GetRasterBand(band)
        if srcband is None:
            continue
        # TODO: check why the "force" doesn't work on GetStatistics but the ComputeStatistics works
        if force:
            s = srcband.ComputeStatistics(0)
            #s = srcband.ComputeRasterMinMax(False)
        else:
            s = srcband.GetStatistics(False, force)
        if stats is None:
            continue
        if math.isnan(s[2]):
            log.warn("polygon is empty! %s " % s)
        else:
            stats.append({"band": band, "min": s[0], "max": s[1], "mean": s[2], "sd": s[3]})


    return stats


def _get_histogram(ds, config=stats_config["histogram"]):
    #log.info("config %s " % config)
    # variables
    # TODO boolean of config value
    force = True if "force" not in config else bool(config["force"])
    buckets = 256 if "buckets" not in config else int(config["buckets"])
    min = None if "min" not in config else int(config["min"])
    max = None if "max" not in config else int(config["max"])
    include_out_of_range = 0 if "include_out_of_range" not in config else int(config["include_out_of_range"])

    # stats
    stats = []
    for band in range(ds.RasterCount):
        band += 1
        # TODO: handle better min max
        if min is None and max is None:
            if force:
                (min, max) = ds.GetRasterBand(band).ComputeRasterMinMax(0)
            else:
                min = ds.GetRasterBand(band).GetMinimum()
                max = ds.GetRasterBand(band).GetMaximum()
        #hist = ds.GetRasterBand(band).GetDefaultHistogram( force = 0 )
        #stats.append({"band": band, "buckets": hist[2], "min": hist[0], "max": hist[1], "values": hist[3]})
        hist = ds.GetRasterBand(band).GetHistogram(buckets=buckets, min=min, max=max, include_out_of_range=include_out_of_range, approx_ok = False )
        stats.append({"band": band, "buckets": buckets, "min": min, "max": max, "values": hist})
    return stats



def get_nodata_value(file_path, band=1):
    try:
        with rasterio.open(file_path) as src:
            return "none" if "nodata" not in src.meta else str(src.meta["nodata"])
    except Exception, e:
        log.error(e)
        raise Exception(e)

# def get_nodata_value(file_path, band=1):
#     ds = gdal.Open(file_path)
#     return ds.GetRasterBand(band).GetNoDataValue()


def get_authority(file_path):
    '''
    Get the authority used by a raster i.e. EPSG:4326
    :param file_path: path to the file
    :return: return the SRID of the raster projection
    '''
    with rasterio.open(file_path) as src:
        log.info(src.meta)
        # if 'init' in src.meta['crs']:
        #     return src.meta['crs']['init']
        # elif 'proj' in src.meta['crs']:
        #     return src.meta['crs']['proj']
        if 'init' in src.meta['crs']:
            return src.meta['crs']['init']
        elif 'proj' in src.meta['crs']:
            # TODO: check if works (find a raster to test it...)
            return "EPSG:" + str(get_epsg_code_from_proj4(src.meta['crs']['proj']))
    return None


def get_srid(file_path):
    '''
    Get the SRID of a raster (i.e. 4326 or 3857 and not EPSG:4326)
    :param file_path: path to the file
    :return: return the SRID of the raster projection
    '''
    proj = get_authority(file_path)
    if ":" in proj:
        return proj.split(":")[1]
    if proj.isdigit():
        return proj
    return None